tyrian/js/src/main/scala/tyrian/Navigation.scala

```diff
@@ -1,42 +1,48 @@
 package tyrian
 
+import cats.effect.kernel.Async
 import org.scalajs.dom.window
 import tyrian.Sub
 import tyrian.facades.HashChangeEvent
 
 import scala.scalajs.js
 import scala.util.control.NonFatal
 
+/** Provides simple routing based on url hash (anchor), such as: `http://mysite.com/#page1` */
 object Navigation:
```

Comment:
We briefly discussed this one privately, but just for the record and any curious readers: Tagless Final support isn't really what I'm after here (I don't think).

Here's an example of a few of those things you mentioned in use as the body of the `init` function:

```scala
val cmds: Cmd[IO, Msg] =
  Cmd.Batch(
    HotReload.bootstrap(hotReloadKey, Model.decode) {
      case Left(msg)    => Msg.Log("Error during hot-reload!: " + msg)
      case Right(model) => Msg.OverwriteModel(model)
    },
    Logger.info(flags.toString),
    Navigation.getLocationHash {
      case Navigation.Result.CurrentHash(hash) => Msg.NavigateTo(Page.fromString(hash))
      case _                                   => Msg.NavigateTo(Page.Page1)
    }
  )

(Model.init, cmds)
```

The thing about the TEA pattern is that you do not do all of your work within some `F` type, as you'd expect to do if you were building a Scala microservice. Most of your code is pretty vanilla 'unsafe' functions that could do anything, and right at the edges (as in the example above) you can describe effects. Each of the main functions that make up a TEA pattern program are as pure as you code them to be. Those edge effects are processed in the background for you by the runtime. As such, an `F` with capabilities would not be passed around, it does not pervade the system, it's only used locally in the `Cmd`/`Sub` implementation itself (and sure, you can use a low power typeclass there by all means). All the end users are going to see is a concrete `IO`.

This would not be true if I changed the API to something sort of like the below, where everything is calculated within an abstract `F` type:

```diff
- def update(msg: Msg, model: Model): (Model, Cmd[IO, Msg]) = ???
+ def update(msg: Msg, model: Model): F[(Model, Cmd[IO, Msg])] = ???
```

People can agree to disagree on whether the TEA pattern is a good idea or not, but that is the pattern. I suspect FRP might be able to cope with Streams and Tagless Final, but it feels a bit much here.

The reason I like the TEA pattern and Tyrian's flavour of it in particular is that while it isn't the most correct and pure form of Scala FP out there, it is very approachable, easy to use, and easy to achieve things in. I'd hate to spoil that for the users with too much in the way of heavy FP. I can imagine that some (I don't mean you) might think that's a bit of an anti-intellectual idea, but I think it's just a design choice. There's something very clean and straightforward about the APIs that I really enjoy.